storage/transfermanager/downloader.go

```diff
@@ -0,0 +1,243 @@
+}
+
+// DownloadObject queues the download of a single object. If it's larger than
+// the specified part size, the download will automatically be broken up into
+// multiple range reads. This will initiate the download but is non-blocking;
+// call Downloader.Results to process the result.
+func (d *Downloader) DownloadObject(ctx context.Context, input *DownloadObjectInput) {
+	input.ctx = ctx
+	d.work <- input
+}
+
+// Download a set of objects to a local path. Downloader will create the needed
+// directory structure locally as the operations progress.
+// This will initiate the download but is non-blocking; wait on Downloader.Results
+// to process the result. Results will be split into individual objects.
+// NOTE: Do not use, DownloadDirectory is not implemented.
+func (d *Downloader) DownloadDirectory(ctx context.Context, input *DownloadDirectoryInput) {
+	d.output <- &DownloadOutput{Bucket: input.Bucket, Err: errors.New("DownloadDirectory is not implemented")}
+	// This does obj listing/os.File calls and then converts to object inputs.
+}
+
+// Waits for all outstanding downloads to complete. The Downloader must not be
+// used to download more objects or directories after this has been called.
+func (d *Downloader) WaitAndClose() error {
+	close(d.work)
+	d.workers.Wait()
+	close(d.output)
+	return nil
+}
+
+// Results returns the iterator for download outputs.
+func (d *Downloader) Results() *DownloadOutputIterator {
+	return &DownloadOutputIterator{
+		output: d.output,
+	}
+}
+
+// downloadWorker continuously processes downloads until the work channel is closed.
+func (d *Downloader) downloadWorker() {
+	d.workers.Add(1)
+	for {
+		input, ok := <-d.work
+		if !ok {
+			break // no more work; exit
+		}
+
+		// TODO: break down the input into smaller pieces if necessary; maybe as follows:
+		// Only request partSize data to begin with. If no error and we haven't finished
+		// reading the object, enqueue the remaining pieces of work and mark in the
+		// out var the amount of shards to wait for.
+		d.output <- input.downloadShard(d.client, d.config.perOperationTimeout)
+	}
+	d.workers.Done()
+}
+
+// NewDownloader creates a new Downloader to add operations to.
+// Choice of transport, etc is configured on the client that's passed in.
+func NewDownloader(c *storage.Client, opts ...TransferManagerOption) (*Downloader, error) {
+	const (
+		chanBufferSize = 1000 // how big is it reasonable to make this?
+		// We should probably expose this as max concurrent ops, because programs can deadlock if calling d.Waitclose before processing it.Next
+	)
+
+	d := &Downloader{
+		client:  c,
+		config:  initTransferManagerConfig(opts...),
+		output:  make(chan *DownloadOutput, chanBufferSize),
+		work:    make(chan *DownloadObjectInput, chanBufferSize),
+		workers: &sync.WaitGroup{},
+	}
+
+	// Start workers in background.
+	for i := 0; i < d.config.numWorkers; i++ {
```

Comment:
Presumably we could optimize this by spinning up workers as needed when there are objects enqueued? Doesn't have to be in this PR though.

Reply:
Sure, though I'm not sure how much that would optimize this by... I guess it depends on the num of workers.

Reply:
Yeah something we can test out later.